packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,123 @@
+import { applyPatches, onPatches, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+import { failure } from "../utils"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposers: Array<() => void>
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot, { generateNewIds: false })
+
+    let wasRS = false
+    const disposeReactionRS = reaction(
+      () => isRootStore(subtreeRoot),
+      isRS => {
+        if (isRS !== wasRS) {
+          wasRS = isRS
+          if (isRS) {
+            registerRootStore(this.subtreeRootDraft)
+          } else {
+            unregisterRootStore(this.subtreeRootDraft)
+          }
+        }
+      },
+      { fireImmediately: true }
+    )
+
+    const disposeOnPatches = onPatches(subtreeRoot, patches =>
+      applyPatches(this.subtreeRootDraft, patches)
+    )
+
+    this.disposers = [disposeReactionRS, disposeOnPatches]
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is called with a draft copy of `node`. Any changes made to the draft
+   * are applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
+    assertTweakedObject(node, "node")
+
+    const path = getParentToChildPath(this.subtreeRoot, node)
+    if (!path) {
+      throw failure("node is not a child of subtreeRoot")
+    }
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value!
+
+    const recorder = patchRecorder(this.subtreeRootDraft)
+    let accept: ReturnType<typeof fn> = false
+    try {
+      accept = fn(draftNode)
+    } finally {
+      recorder.dispose()
+      if (accept) {
+        for (const event of recorder.events) {
```

Comment:
looks good, but bear in mind that the transpilation of for-of is kind of slow. if you want it to be more performant then use something like
```
const len = arr.length
for (let i = 0; i < len; i++) {....}
```

Reply:
Uh, I didn't know about the transpilation to slow code. I changed it as you suggested in 78c2d3dd0b66100c57ff3bcbaecf79f3927a02a4.